THNN.lua

```diff
@@ -0,0 +1,133 @@
+local ffi = require 'ffi'
+
+local THNN = {}
+
+local generic_THNN_h = [[
+TH_API void THNN_(Abs_updateOutput)(
+          THNNState *state,
+          THTensor *input,
+          THTensor *output);
+TH_API void THNN_(Abs_updateGradInput)(
+          THNNState *state,
+          THTensor *input,
+          THTensor *gradOutput,
+          THTensor *gradInput);
+          
+TH_API void THNN_(AbsCriterion_updateOutput)(
+          THNNState *state,
+          THTensor *input,
+          THTensor *target,
+          real *output,
+          bool sizeAverage);
+TH_API void THNN_(AbsCriterion_updateGradInput)(
+          THNNState *state,
+          THTensor *input,
+          THTensor *target,
+          THTensor *gradInput,
+          bool sizeAverage);
+]]
+
+-- THGenerator struct declaration copied from torch7/lib/TH/THRandom.h
+local base_declarations = [[
+typedef void THNNState;
+
+typedef struct {
+  unsigned long the_initial_seed;
+  int left;
+  int seeded;
+  unsigned long next;
+  unsigned long state[624]; /* the array for the state vector 624 = _MERSENNE_STATE_N  */
+  double normal_x;
+  double normal_y;
+  double normal_rho;
+  int normal_is_valid;
+} THGenerator;
+]]
+
+ffi.cdef(base_declarations)
+
+-- expand macros, allow to use original lines from lib/THNN/generic/THNN.h
+local preprocessed = string.gsub(generic_THNN_h, 'TH_API void THNN_%(([%a%d_]+)%)', 'void THNN_TYPE%1')
+
+local replacements = 
+{
+  { ['TYPE'] = 'Double', ['real'] = 'double', ['THTensor'] = 'THDoubleTensor', ['THIndexTensor'] = 'THLongTensor' },
+  { ['TYPE'] = 'Float',  ['real'] = 'float',  ['THTensor'] = 'THFloatTensor',  ['THIndexTensor'] = 'THLongTensor' }
+}
+
+for i=1,#replacements do
+  local r = replacements[i]
+  local s = preprocessed
+  for k,v in pairs(r) do
+    s = string.gsub(s, k, v)
+  end
+  ffi.cdef(s)
+end
+
+THNN.NULL = ffi.NULL or nil
+
+function THNN.getState()
+  return ffi.NULL or nil
+end
+
+function THNN.optionalTensor(t)
+  return t and t:cdata() or THNN.NULL
+end
+
+local ok,result
+if ffi.os == "OSX" then
```

Comment:
this whole section around ffi.load should be:
(1) made this way: https://github.com/torch/tds/blob/735f6cd37e7e3385c8eb229d1c38969f07436542/env.lua#L5
     So, THNN will be installed into cpath, and we know exactly where it will be in all OSes.
(2) No pcall around it. It should be a hard failure.